accumulate.go

```diff
@@ -31,66 +40,121 @@ namespace are stored.
 */
-func accumulateBucketMetrics(ctx context.Context, date time.Time, cloudscaleClient *cloudscale.Client) (map[AccumulateKey]uint64, error) {
+func accumulateBucketMetrics(ctx context.Context, date time.Time, cloudscaleClient *cloudscale.Client, k8sclient client.Client) (map[AccumulateKey]uint64, error) {
 	bucketMetricsRequest := cloudscale.BucketMetricsRequest{Start: date, End: date}
 	bucketMetrics, err := cloudscaleClient.Metrics.GetBucketMetrics(ctx, &bucketMetricsRequest)
 	if err != nil {
 		return nil, err
 	}
 
+	nsOrganizations, err := fetchNamespaces(ctx, k8sclient, client.HasLabels{organizationLabel})
+	if err != nil {
+		return nil, err
+	}
+
+	buckets, err := fetchBuckets(ctx, k8sclient, client.HasLabels{namespaceLabel})
+	if err != nil {
+		return nil, err
+	}
+
 	accumulated := make(map[AccumulateKey]uint64)
 
 	for _, bucketMetricsData := range bucketMetrics.Data {
-		objectsUser, err := cloudscaleClient.ObjectsUsers.Get(ctx, bucketMetricsData.Subject.ObjectsUserID)
-		if err != nil || objectsUser == nil {
-			fmt.Fprintf(os.Stderr, "WARNING: Cannot sync bucket %s, objects user %s not found\n", bucketMetricsData.Subject.BucketName, bucketMetricsData.Subject.ObjectsUserID)
+		name := bucketMetricsData.Subject.BucketName
+		ns, ok := buckets[name]
+		if !ok {
+			fmt.Fprintf(os.Stderr, "WARNING: Cannot sync bucket, bucket resource %q not found\n", name)
 			continue
 		}
-		err = accumulateBucketMetricsForObjectsUser(accumulated, bucketMetricsData, objectsUser)
+		organization, ok := nsOrganizations[ns]
+		if !ok {
+			fmt.Fprintf(os.Stderr, "WARNING: Cannot sync bucket, namespace %q not found in map\n", ns)
+			continue
+		}
+
+		err = accumulateBucketMetricsForObjectsUser(accumulated, bucketMetricsData, organization, ns)
 		if err != nil {
-			fmt.Fprintf(os.Stderr, "WARNING: Cannot sync bucket %s: %v\n", bucketMetricsData.Subject.BucketName, err)
+			fmt.Fprintf(os.Stderr, "WARNING: Cannot sync bucket %s: %v\n", name, err)
 			continue
 		}
 	}
 
 	return accumulated, nil
 }
 
-func accumulateBucketMetricsForObjectsUser(accumulated map[AccumulateKey]uint64, bucketMetricsData cloudscale.BucketMetricsData, objectsUser *cloudscale.ObjectsUser) error {
-	if len(bucketMetricsData.TimeSeries) != 1 {
-		return fmt.Errorf("There must be exactly one metrics data point, found %d", len(bucketMetricsData.TimeSeries))
+func fetchBuckets(ctx context.Context, k8sclient client.Client, opts ...client.ListOption) (map[string]string, error) {
+	buckets := &cloudscalev1.BucketList{}
+	if err := k8sclient.List(ctx, buckets, opts...); err != nil {
+		return nil, fmt.Errorf("bucket list: %w", err)
+	}
+
+	bucketNS := map[string]string{}
+	for _, b := range buckets.Items {
+		bucketNS[b.Spec.ForProvider.BucketName] = b.Labels[namespaceLabel]
+	}
+
+	if buckets.Continue != "" {
+		m, err := fetchBuckets(ctx, k8sclient, client.Continue(buckets.Continue))
```

Comment:
Can you explain why are you using this function recursively? I noticed you did the same operation for namespaces as well. From my limited point of view I don't see you are fetching a limited number of objects (buckets and namespaces) from K8s.

Reply:
it was more of a defensive programming point of view - I'm not 100% certain whether it's never returning a paginated list if you don't specify a limit. If it's always, regardless of the amount of items, returning all resources, I'm more than fine with removing it. The documentation I read weren't entirely clear on that though I'm leaning towards that if no limit is specified, no pagination is done at all. Do you know how it behaves exactly?

Reply:
> Do you know how it behaves exactly?

Unfortunately I'm not familiar with it and I cannot give you an answer. I am inclined to believe that if you don't specify a limit the server will not return a paginated list and for this reason I would simplify the function. Also you could change it and make sure it returns a paginated list (to justify the complexity) though again not sure how useful it would be for the number of items we might have (100-200 for namespaces and and maybe a max of 100 instances of objectbuckets?). @Kidswiss what's your take on that?

Reply:
I've found this page here, which describes the concept of chunking with the continue parameter: https://kubernetes.io/docs/reference/using-api/api-concepts/#retrieving-large-results-sets-in-chunks
AFAIU, this is only a problem with really large datasets, e.g. thousands of Pods. Since the JSON of Pods is significantly larger than Namespaces or Buckets, I'd say we don't need to use pagination/continue at this time and just list everything. Since this is a job in the background run at most a few times per day, it doesn't matter if data loading takes a few seconds.

Reply:
:+1: agreed & removed that bit from both functions.